Node tree rewrite

kvark opened the pull request:
Fixes #44

Core of the change is switching relationships from "child->parent" to "parent->first_child->sibling", which gives us the following benefits:
  - clean children lifetimes, as expected by users
    - no need to store all the meshes returned by gltf/obj loaders: perhaps, we can follow-up with a PR that allows querying meshes by name from their parents, so that we don't even return all the meshes directly in the loader API
  - nice `Group::add(&something)`, like Three-js does
  - make `Scene` truly unique and node-less
  - no need for unique scene IDs
    ~~- technically, objects can now be present in multiple scenes at once~~
  - immutable node data upon traversal (!):
    - we just compute the world visibility/transform on the fly
    - thus, no need for a separate `update_transforms` step
    - uses pretty iterators ;)
    - as a downside, not as straightforward to reflect the world info to the user (see TODO)

Has other changes making things nicer:
  - refactored frame render code
  - no `&mut self` on objects
  - idiomatic hub indexing
  - safer group/camera/text/audio spawning (internal)
  - more robust message handling (panic on unprocessed/unexpected messages)
  - can debug print the `Operation`

TODO:
- [x] check/fix docs
- [x] compute world transform/visibility for the user if needed (is it needed?)
- [x] provide API to detach from parent
- [x] reviews!
- [x] profile the cost of recomputing the world transforms

vitvakatu commented on src/hub.rs:
```diff
@@ -185,24 +148,55 @@ impl Hub {
+        for (child_ptr, sibling) in deferred_sibling_updates {
+            let child = &mut self.nodes[&child_ptr];
+            if child.next_sibling.is_some() {
+                error!("Attaching a child that still has an old parent, discarding siblings");
```
How we can change parent now?

kvark replied:
Oh right, I forgot about this :) Now added to TODO list for this PR.
Basically, we can either merge the old siblings with the new ones (like `Scene::add` does), or require the user to call `Group::remove(&item)` before adding to another group. Either way, the `remove` needs to be added.

alteous commented on src/hub.rs:
```diff
@@ -185,24 +148,55 @@ impl Hub {
+                Operation::AddChild(child_ptr) => match node.sub_node {
+                    SubNode::Group { ref mut first_child } => {
+                        let sibling = mem::replace(first_child, Some(child_ptr.clone()));
+                        deferred_sibling_updates.push((child_ptr, sibling));
```
Interesting.

kvark replied:
borrow checker...

alteous commented on src/hub.rs:
```diff
@@ -240,40 +234,83 @@ impl Hub {
         }
     }
+
+    pub(crate) fn walk(&self, base: &Option<NodePointer>) -> TreeWalker {
```
👍

alteous commented on src/node.rs:
```diff
@@ -72,28 +74,10 @@ impl From<TransformInternal> for Transform {
-    pub world_visible: bool,
+    /// Relative to parent transform.
+    pub transform: Transform,
     /// Material in case this `Node` has it.
```
Was this ever useful?

kvark replied:
wait, why is `Material` here? O_O

alteous replied:
Why not?!

kvark replied:
Ah, this is user-exposed node, not the internal one.
Yeah, let it be. The whole reflection thing (with `sync`) I'd consider non-idiomatic, so we shouldn't be too eager to optimize it.

alteous commented on src/render/mod.rs:
```diff
@@ -654,13 +653,9 @@ impl Renderer {
-            };
-            let mx_view = Matrix4::from(w.inverse_transform().unwrap());
+            let node = &hub[&camera];
+            let world_transform = node.transform; //TODO!!!
```
What's this for?

kvark replied:
TODO: compute the world transform for the camera, which is expected here

alteous commented on src/scene.rs:
```diff
@@ -4,15 +4,13 @@ use object;
 use std::sync::MutexGuard;
 
-/// Unique identifier for a scene.
-pub type Uid = usize;
```
👍

kvark commented:
Alright, I rebased the code and implemented the missing bits.
Only the profiling is left, plus maybe a few more looks, esp in the new code.
Note: unfortunately, there is a bit of duplication atm between Group reacting to events and Scene implementation. I'm sure we can make it nicer.

kvark commented:
Here are the profiling results (note: I only used `callgrind_annotate`, so don't have that much insight at the moment):
  1. `descend` does show up near the top of the profile
  2. cause likely being large amount of copying (and a bit of `stack` reallocation). This should be partially solved by future improvements in Rust itself, some possible workarounds from our side, or caching the transforms at least.
  3. `group` examples with 20K instances runs OK on my U-series CPU, so I think we are fine for now.

Looks like I addressed all the action items from my side. Waiting for more reviews now (@alteous , @vitvakatu ).

kvark commented:
I believe in the long term our architecture should work like this:
- `Renderer` only walks the tree once, for each node
  - figures out what render bins does an object fall into (e.g.: main screen, shadow for light X, off-screen buffer, etc)
  - builds the GPU instance blocks for each bin (separately) and appends them to lists to be directly uploaded to GPU constant buffer memory
  - writes down GPU commands in the appropriate command buffers

TL;DR: there is no strong reason for `Renderer` to walk the tree multiple times. We can make rendering much faster on the CPU side, and tree traversal is not (going to be) a bottleneck.

alteous commented on examples/group.rs:
```diff
@@ -102,28 +102,41 @@ fn create_cubes(
-const SPEEDS: [f32; 6] = [0.7, -1.0, 1.3, -1.6, 1.9, -2.2];
+struct LevelDesc {
+    color: three::Color,
+    speed: f32,
```
What is the unit of this? I assume radians/second?

kvark replied:
correct

vitvakatu commented on Cargo.toml:
```diff
@@ -28,7 +31,7 @@ cgmath = { version = "0.15", features = ["mint"] }
 froggy = "0.4.4"
 genmesh = "0.5"
-gfx = "0.16"
+gfx = "0.16.3"
```
Why?

kvark replied:
Because your instancing changes require it

vitvakatu replied:
I'm not a fan of strict version rules in Cargo. It will use the latest available version anyway.

kvark replied:
I think your position is in the minority within the community then. Latest will be used on fresh checkout, if we don't check in `Cargo.lock`. But people who already have `Cargo.lock` will get breakage if you don't write the exact version in the `Cargo.toml`. Another case is - some crate that depends on you may say `gfx = "=0.16.2"` and it will try to build but will fail even after `cargo update`.

TL;DR: Cargo versioning rules are here to stay and be respected

kvark replied:
Also, there is a proposal (which I supported) about making `cargo publish` checking out all the **minimal** compatible versions (as opposed to latest). Think about it ;)

vitvakatu replied:
Okay, you convicted me

alteous commented on src/hub.rs:
```diff
@@ -240,40 +272,94 @@ impl Hub {
+        walker
+    }
+
+    pub(crate) fn walk(&self, base: &Option<NodePointer>) -> TreeWalker {
```
Nit: Perhaps `walk_visible`?

vitvakatu replied:
As in froggy, it's better to use shorter names.

kvark replied:
it's for internal use only anyhow. Most of the time we are only concerned about visible nodes

vitvakatu replied:
Yes, I remember your strategy :smile:
![alt](https://screenshots.firefoxusercontent.com/images/9194b8ff-3e7c-4ce6-ae9a-5fe0647cc19d.png)

vitvakatu commented on examples/shapes.rs:
```diff
@@ -7,34 +7,40 @@ use three::Object;
 
-    let mut mline = {
+    // test removal from scene
+    win.scene.remove(&mcyl);
```
Is it supposed to be here?

kvark replied:
yes - it's a small test for object removal. May not be the best place, but better than nothing

vitvakatu commented:
bors r+

kvark commented:
bors r-
Almost forgot to fix the text!!!

kvark commented:
bors r=vitvakatu
